cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/DefaultCheckpointManager.java

```diff
@@ -81,50 +83,62 @@ public void saveCheckpoints(final Map<Integer, ? extends Checkpoint> checkpoints
       return;
     }
 
-    transactional.execute(new TxRunnable() {
-      @Override
-      public void run(DatasetContext context) throws Exception {
-        Table table = getCheckpointTable(context);
-        for (Map.Entry<Integer, ? extends Checkpoint> entry : checkpoints.entrySet()) {
-          byte[] key = Bytes.add(rowKeyPrefix, Bytes.toBytes(entry.getKey()));
-          Checkpoint checkpoint = entry.getValue();
-          table.put(key, OFFSET_COL_NAME, Bytes.toBytes(checkpoint.getNextOffset()));
-          table.put(key, MAX_TIME_COL_NAME, Bytes.toBytes(checkpoint.getMaxEventTime()));
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          Table table = getCheckpointTable(context);
+          for (Map.Entry<Integer, ? extends Checkpoint> entry : checkpoints.entrySet()) {
+            byte[] key = Bytes.add(rowKeyPrefix, Bytes.toBytes(entry.getKey()));
+            Checkpoint checkpoint = entry.getValue();
+            table.put(key, OFFSET_COL_NAME, Bytes.toBytes(checkpoint.getNextOffset()));
+            table.put(key, MAX_TIME_COL_NAME, Bytes.toBytes(checkpoint.getMaxEventTime()));
+          }
         }
-      }
-    });
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e, ServiceUnavailableException.class);
```

Comment:
The `TransactionFailureException` wraps the underlying cause. The `Transactions.propagate` is to unwrap it. Unwrapping and servicing the `ServiceUnavilableException` is necessary for retry.